### ISSUE TITLE
Bug: lambda *args: argument type is partially unknow

### ISSUE THREAD

**Issue by ZashIn** [microsoft/pyright#6222]
Lambdas with `*args` are wrongly reported with partially unknown types (`reportUnknownMemberType`).

A declared function does not report any issues, as expected.
Mypy handles this correctly.

```py
from typing import Callable, Any

def test(fn: Callable[[int, int, int], int]) -> int:
    return fn(1, 2, 3)

def callback(a: int, *_: Any) -> int:
    return a

# expected: no errors
test(lambda a, *_: a)  # actual: reportUnknownMemberType
test(callback)

# expected error: wrong return type
test(lambda a, *_: str(a))  # type: ignore [reportUnknownMemberType]
# actual: no error
```
pyright output
```
test.py
  test.py:10:6 - error: Argument type is partially unknown
    Argument corresponds to parameter "fn" in function "test"
    Argument type is "(a: Unknown, *_: Unknown) -> Unknown" (reportUnknownArgumentType)
  test.py:10:13 - error: Type of parameter "a" is unknown (reportUnknownLambdaType)
  test.py:10:17 - error: Type of parameter "_" is partially unknown (reportUnknownLambdaType)
  test.py:10:20 - error: Return type of lambda is unknown (reportUnknownLambdaType)
4 errors, 0 warnings, 0 informations
```
mypy output:
```
test.py:14: error: Argument 1 to "test" has incompatible type "Callable[[int, int, int], str]"; expected "Callable[[int, int, int], int]"  [arg-type]
test.py:14: note: Error code "arg-type" not covered by "type: ignore" comment
```

**Comment by erictraut** [comment]
I took this opportunity to generalize pyright's type inference logic for lambdas. It now handles a broader range of cases, including the one used by the code above. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.334, which I just published. It will also be included in a future release of pylance.

**Comment by ZashIn** [comment]
Thanks for the fast fix @erictraut. Sadly this seems to work only for args with same type, but not for mixed:
```py
class A: pass
class B(A): pass

def test(fn: Callable[[A, B], None]): pass
def test2(fn: Callable[[B, A], None]): pass

test(lambda a, b: None)
test2(lambda a, b: None)
test(lambda *_: None)
test2(lambda *_: None)  # error: Argument of type "(*_: B) -> None" cannot be assigned to parameter "fn" of type "(B, A) -> None" in function "test2"
```
So the first `Callable` parameter type is assumed for all `*args`?
I understand the difficulty for a static type checker to determine the types, but the results in this case are a bit confusing / inconsistent.

One could argue that this is an edge case, but using such a pattern for callbacks, which do not use some/any arguments is convenient (and imo cleaner), since the actual signature does not matter, instead of adding unused parameters to the lambda.

**Comment by erictraut** [comment]
Yes, my solution didn't account for all possibilities. Function signatures in Python are very complicated (overloads, positional vs. keyword, `*args`, `*kwargs`, ParamSpecs, TypeVarTuples, etc.). Determining whether one function signature is compatible with another function signature involves dealing with the combination of all all of the above complexities for both signatures.

Someone else noted the regression that my fix caused in [this bug report](https://github.com/microsoft/pyright/issues/6308). A fix for this will appear in the next release.
